mongoimport/common.go

```diff
@@ -552,47 +587,168 @@ func validateFields(fields []string, useArrayIndexFields bool) error {
 		if strings.Contains(field, "..") {
 			return fmt.Errorf("field '%v' cannot contain consecutive '.' characters", field)
 		}
-		// NOTE: since fields is sorted, this check ensures that no field
-		// is incompatible with another one that occurs further down the list.
-		// meant to prevent cases where we have fields like "a" and "a.c"
-
-		// TODO: Add a check here for incompatible fields:
-		// a.b and a.0
-		for _, latterField := range fieldsCopy[index+1:] {
-			// NOTE: this means we will not support imports that have fields that
-			// include e.g. a, a.b
-			if strings.HasPrefix(latterField, field+".") {
-				return fmt.Errorf("fields '%v' and '%v' are incompatible", field, latterField)
+
+		if index+1 < len(fields) {
+			// For each field except the last, we check that the next field does not
+			// contain the current field as a prefix. Since fields is sorted, this checks
+			// case (3) and case (4).
+			nextField := fields[index+1].field
+
+			if strings.HasPrefix(nextField, field+".") {
+				return fmt.Errorf("fields '%v' and '%v' are incompatible", field, nextField)
 			}
-			// NOTE: this means we will not support imports that have fields like
-			// a, a - since this is invalid in MongoDB
-			if field == latterField {
-				return fmt.Errorf("fields cannot be identical: '%v' and '%v'", field, latterField)
+
+			// This checks case (2).
+			if field == nextField {
+				return fmt.Errorf("fields cannot be identical: '%v' and '%v'", field, nextField)
 			}
 		}
 
 		if useArrayIndexFields {
-			// Check for incompatible fields: a.n.c, a.b.c (where n = 0,1,2,...)
-			// NOTE: We must scan the whole fieldsCopy array since some fields could
-			// come before fields with numbers if key parts start with symbols
-			fieldParts := strings.Split(field, ".")
-			for _, otherField := range fieldsCopy {
-				otherFieldParts := strings.Split(otherField, ".")
-				numOtherFieldParts := len(otherFieldParts)
-				for i, part := range fieldParts {
-					if i == numOtherFieldParts {
-						// Out of bounds of otherFieldParts
-						break
-					}
-					if otherFieldParts[i] != part {
-						if !(isNatNum(otherFieldParts[i]) == isNatNum(part)) {
-							return fmt.Errorf("fields '%v' and '%v' are incompatible", field, otherField)
-						}
-						break
-					}
+			var nextFieldInfo *FieldInfo
```

Comment:
I don't think you need pointers for these: FieldInfo is a small, cheap structure to copy, as the slice is just a pointer-type anyway.  I think the code will be clearer, as it won't make the reader wonder why you're passing by reference and if something is actually depending on mutability.